### ISSUE TITLE
[BUG] Adjoint of `v[i] = v[i] + x` on local vectors/matrices appears silently wrong under the default `enable_vector_component_overwrites=False` (no warning; compounds exponentially across taped launches)

### ISSUE THREAD

**Issue by Ramzee-S** [NVIDIA/warp#1697]
## Summary

First of all: we are not sure whether this is an actual bug or a
misunderstanding of the intended contract on our side — the existence of
`wp.config.enable_vector_component_overwrites` suggests that multiple
writes to a component may simply be unsupported by default, and that we
were (unknowingly) outside that contract. We thought it was worth filing
anyway, for clarification if nothing else, because the failure mode is
unusually hard to catch from the user side: the forward result is
bit-correct, no warning is emitted at any log level, and the gradient
error compounds exponentially across taped launches — so in a multi-step
differentiable simulation you get plausible-looking gradients at short
horizons and NaN at longer ones, with nothing pointing at the offending
line.

How we ran into it: in a differentiable explicit FEM solver, an unrolled
4-Gauss-point accumulation into a local matrix made 200-step gradients go
`3.6e-4 → 6.8e+2 → 2.7e+12 → NaN` at 5/10/20/40 steps while the forward
physics stayed exactly right. It took an FD cross-check gate to notice
anything was wrong at all.

## Reproduction (full script attached)

```python
mat22 = wp.types.matrix(shape=(2, 2), dtype=wp.float64)

@wp.kernel
def rmw_general(x: wp.array(dtype=wp.float64), y: wp.array(dtype=wp.float64)):
    tid = wp.tid()
    m = mat22()
    for i in range(4):                     # static bound -> unrolled
        m[0, 0] = m[0, 0] + x[tid] * wp.float64(1.0 + float(i))
    y[tid] = m[0, 0]
```

`y = (1+2+3+4)·x`, so `dy/dx = 10`. What we observe (identical on CPU and
`cuda:0`, fp32 and fp64, warp 1.15.0 and 1.16.0.dev20260724):

```
mat  general RMW  m[0,0] = m[0,0]+x: forward= 10.0  adjoint=  26.0   <- unexpected
mat  augmented    m[0,0] += x      : forward= 10.0  adjoint=  10.0
mat  whole-add    m = m + mat22(..): forward= 10.0  adjoint=  10.0
vec  general RMW  v[0] = v[0]+x    : forward= 10.0  adjoint=  26.0   <- unexpected
     scalar       s = s + x        : forward= 10.0  adjoint=  10.0
```

The deviation is structured, not noise: write k of an n-write RMW chain
appears with adjoint weight 2^(n−k), so `26 = 4·2⁰ + 3·2¹ + 2·2² + 1·2³`.

Chaining an identity-forward step built from four `0.25`-quarters over N
taped launches gives an adjoint of exactly `3.75^N` (true value 1) — this
is the exponential-compounding aspect:

```
N= 1: forward=1.000  adjoint=        3.75
N= 4: forward=1.000  adjoint=       197.8
N= 8: forward=1.000  adjoint=   3.911e+04
N=16: forward=1.000  adjoint=   1.529e+09
```

## Our reading of what happens (please correct us if wrong)

`adj_assign_inplace` (native/vec.h; the matrix version in native/mat.h
looks structurally identical) propagates the overwritten component's
adjoint but leaves the slot in place:

```cpp
adj_value += adj_v[idx];      // adj_v[idx] is not reset
```

If we understand overwrite semantics correctly, the component's previous
value is dead after the assignment, so its adjoint slot would need to be
reset after propagating into `adj_value`; otherwise, when the backward
pass later reaches the RHS *read* of the same component, the stale slot is
counted a second time — one doubling per chained write, which matches the
observed `2^(n−k)` weighting and the `3.75^N` chain law exactly. But we
may well be missing an invariant of the lowering that makes this the
intended behavior.

Three related observations:

- `wp.config.enable_vector_component_overwrites = True` makes every case
  above exact (we verified this), so a supported path clearly exists.
- The augmented spelling `m[0,0] += x` (→ `add_inplace`) is exact, while
  the semantically identical `m[0,0] = m[0,0] + x` is not — which is what
  made this hard for us to anticipate.
- We could not find a warning for this pattern at any log level (the
  mutating-reference path warns at LOG_DEBUG; this one seems silent).

## Questions / suggestions

Whatever the intended semantics are, we'd be grateful for clarification —
and depending on the answer, perhaps one of these would be worth doing:

1. If this is simply a missing reset: would
   `adj_value += adj_v[idx]; adj_v[idx] = Type(0);` be a viable fix for
   the default mode? (We're aware you know the lowering invariants far
   better than we do — happy to test a patch against our FEM gradient
   gates if useful.)
2. If write-once-per-component is the intended default contract: could a
   codegen diagnostic fire when a local composite's component is assigned
   more than once (or RMW-assigned) in a differentiable kernel with the
   flag off? The silent-with-correct-forward failure mode is the painful
   part; even a LOG_DEBUG warning like the mutating-reference one would
   have saved us a day.
3. Docs: the `enable_vector_component_overwrites` docstring currently
   mentions only "Allow multiple writes… may significantly increase
   kernel compilation time" — a sentence noting that the default produces
   incorrect gradients for multiple component writes (with a pointer in
   the differentiability guide) would let users find this by search the
   first time it bites them.

## System Information

- Warp 1.15.0 (pip) and 1.16.0.dev20260724 — identical behavior
- Python 3.12.12, Ubuntu (kernel 6.5.0-28), x86_64
- Devices: CPU and NVIDIA RTX 4000 SFF Ada (sm_89), CUDA toolkit 12.9,
  driver 12.4 — identical behavior on both
- fp32 and fp64 — identical behavior (attached repro uses fp64)

The full repro is below (self-contained, deterministic, integer-exact
expected values; flipping the commented flag line shows
`enable_vector_component_overwrites=True` fixing all cases).

Thanks for taking a look — and apologies in advance if this turns out to
be us holding the API wrong; a pointer to the relevant docs would fully
resolve it in that case.

*(Context: same project as #1693 and #1694.)*

<details>
<summary><b>warp_rmw_adjoint_repro.py</b> (click to expand)</summary>

```python
"""Standalone repro: silently wrong adjoints for vector/matrix component
read-modify-write (RMW) in warp kernels under the DEFAULT
``wp.config.enable_vector_component_overwrites = False``.

Findings (warp 1.15.0 and 1.16.0.dev20260724, CPU and CUDA identical,
fp32 and fp64):

1. ``m[0,0] = m[0,0] + x`` chains (general-form component RMW on a LOCAL
   matrix/vector) produce adjoints with each write double-counted per
   subsequent write: write k of an n-write chain gets weight 2^(n-k).
   Forward is bit-correct. Root cause: ``adj_assign_inplace`` in
   native/vec.h / native/mat.h does ``adj_value += adj_v[idx];`` WITHOUT
   zeroing ``adj_v[idx]`` — the overwritten component's stale adjoint is
   then re-counted at the RHS read of the same component.
2. Chained over N taped kernel launches the wrong per-launch gain
   compounds geometrically (here exactly 3.75^N) — in a multi-step
   simulation this is an exponentially exploding gradient with a correct
   forward (NaN within tens of steps).
3. The AUGMENTED form ``m[0,0] += x`` lowers to ``add_inplace`` and is
   EXACT — two semantically identical spellings differ.
4. ``wp.config.enable_vector_component_overwrites = True`` (assign_copy /
   SSA lowering) fixes the general form completely (verified: all
   adjoints exact). Its docstring says only "Allow multiple writes...may
   significantly increase kernel compilation time" — not that the default
   silently produces wrong gradients.
5. No warning is emitted at any log level for the broken pattern (the
   reference-mutation path warns at LOG_DEBUG; this path does not).
6. Plain scalar reassignment (``s = s + x``) is exact.

Our current workaround: accumulate into local composites by WHOLE-VALUE
adds (``m = m + mat22(...)``) — correct under the default flag, no
compile-time cost, no reliance on lowering subtleties.

Run: python warp_rmw_adjoint_repro.py   (numbers are integer-exact)
"""

import warp as wp

# flip to True to verify the documented workaround fixes everything:
# wp.config.enable_vector_component_overwrites = True
wp.init()
DEV = "cpu"

mat22 = wp.types.matrix(shape=(2, 2), dtype=wp.float64)

# true dy/dx = 1+2+3+4 = 10 in every single-call case below


@wp.kernel
def rmw_general(x: wp.array(dtype=wp.float64), y: wp.array(dtype=wp.float64)):
    tid = wp.tid()
    m = mat22()
    for i in range(4):  # static bound -> unrolled
        m[0, 0] = m[0, 0] + x[tid] * wp.float64(1.0 + float(i))
    y[tid] = m[0, 0]


@wp.kernel
def rmw_augmented(x: wp.array(dtype=wp.float64), y: wp.array(dtype=wp.float64)):
    tid = wp.tid()
    m = mat22()
    for i in range(4):
        m[0, 0] += x[tid] * wp.float64(1.0 + float(i))
    y[tid] = m[0, 0]


@wp.kernel
def whole_add(x: wp.array(dtype=wp.float64), y: wp.array(dtype=wp.float64)):
    tid = wp.tid()
    m = mat22()
    for i in range(4):
        m = m + mat22(
            x[tid] * wp.float64(1.0 + float(i)), wp.float64(0.0),
            wp.float64(0.0), wp.float64(0.0),
        )
    y[tid] = m[0, 0]


vec2d = wp.types.vector(length=2, dtype=wp.float64)


@wp.kernel
def rmw_vec(x: wp.array(dtype=wp.float64), y: wp.array(dtype=wp.float64)):
    tid = wp.tid()
    v = vec2d()
    for i in range(4):
        v[0] = v[0] + x[tid] * wp.float64(1.0 + float(i))
    y[tid] = v[0]


@wp.kernel
def rmw_scalar(x: wp.array(dtype=wp.float64), y: wp.array(dtype=wp.float64)):
    tid = wp.tid()
    s = wp.float64(0.0)
    for i in range(4):
        s = s + x[tid] * wp.float64(1.0 + float(i))
    y[tid] = s


# identity-forward chain step (four 0.25-quarters): per-step adjoint gain
# under the bug = 0.25*(1+2+4+8) = 3.75, so N steps give exactly 3.75^N
@wp.kernel
def rmw_chain_step(x: wp.array(dtype=wp.float64), y: wp.array(dtype=wp.float64)):
    tid = wp.tid()
    m = mat22()
    for i in range(4):
        m[0, 0] = m[0, 0] + x[tid] * wp.float64(0.25)
    y[tid] = m[0, 0]


def one_call_adjoint(kernel, dev=DEV):
    x = wp.array([1.0], dtype=wp.float64, device=dev, requires_grad=True)
    y = wp.zeros(1, dtype=wp.float64, device=dev, requires_grad=True)
    tape = wp.Tape()
    with tape:
        wp.launch(kernel, dim=1, inputs=[x], outputs=[y], device=dev)
    tape.backward(grads={y: wp.array([1.0], dtype=wp.float64, device=dev)})
    return float(y.numpy()[0]), float(x.grad.numpy()[0])


def chain_adjoint(kernel, nsteps):
    bufs = [wp.array([1.0], dtype=wp.float64, device=DEV, requires_grad=True)]
    tape = wp.Tape()
    with tape:
        for _ in range(nsteps):
            nxt = wp.zeros(1, dtype=wp.float64, device=DEV, requires_grad=True)
            wp.launch(kernel, dim=1, inputs=[bufs[-1]], outputs=[nxt], device=DEV)
            bufs.append(nxt)
    tape.backward(grads={bufs[-1]: wp.array([1.0], dtype=wp.float64, device=DEV)})
    return float(bufs[-1].numpy()[0]), float(bufs[0].grad.numpy()[0])


print(f"warp {wp.config.version}, "
      f"enable_vector_component_overwrites={wp.config.enable_vector_component_overwrites}")

print("\n-- single call, true dy/dx = 10 in every row --")
for name, k in (
    ("mat  general RMW  m[0,0] = m[0,0]+x", rmw_general),
    ("mat  augmented    m[0,0] += x      ", rmw_augmented),
    ("mat  whole-add    m = m + mat22(..)", whole_add),
    ("vec  general RMW  v[0] = v[0]+x    ", rmw_vec),
    ("     scalar       s = s + x        ", rmw_scalar),
):
    y, g = one_call_adjoint(k)
    print(f"{name}: forward={y:5.1f}  adjoint={g:6.1f}")

print("\n-- N-step identity chain (general RMW): forward = 1, true adjoint = 1 --")
for n in (1, 2, 4, 8, 16):
    y, g = chain_adjoint(rmw_chain_step, n)
    print(f"N={n:2d}: forward={y:.3f}  adjoint={g:12.4g}   (3.75^N = {3.75**n:.4g})")

if wp.get_cuda_device_count() > 0:
    print("\n-- CUDA (same single-call cases) --")
    for name, k in (("mat general RMW", rmw_general), ("vec general RMW", rmw_vec),
                    ("mat augmented  ", rmw_augmented)):
        y, g = one_call_adjoint(k, "cuda:0")
        print(f"cuda:0 {name}: forward={y:5.1f}  adjoint={g:6.1f}")
```

</details>

**Comment by shi-eric** [comment]
From the [differentiability guide](https://nvidia.github.io/warp/user_guide/differentiability.html):

> Within a kernel, assigning a value to a locally defined vector, matrix, or quaternion component is differentiable, with one important caveat: each component may only be assigned a value once (not including default initialization). Each component may then be safely updated with in-place addition or subtraction (`+=` or `-=`) operations, but direct re-assignment (`=`) will invalidate gradient computations related to the vector, matrix, or quaternion.

This is the restriction reached by the reproducer. The static loop directly assigns `m[0, 0]` four times, so `m[0, 0] = m[0, 0] + x` is not differentiable under the default configuration. The `m[0, 0] += x` form is supported. Code that needs direct component reassignment can set `wp.config.enable_vector_component_overwrites = True`, which selects the copy-based lowering at the cost of longer compilation times.

Clearing the component's adjoint inside `adj_assign_inplace()` would fix this particular linear RMW example, but it would not be a general solution. Direct reassignment can also destroy earlier composite values needed to differentiate nonlinear operations between assignments. The copy-based path preserves those values.

The lack of a diagnostic is a fair point, though. The current debug mode, debug logging, and autograd array-access verifier do not catch this because the matrix is a local value rather than a kernel array argument.

@daedalus5, could you open a separate issue for an opt-in diagnostic that detects repeated direct assignments to local vector, matrix, quaternion, and transform components? It should be disabled by default and only apply when backward code generation is enabled, so forward-only simulations do not get irrelevant warnings. When enabled, it could report the source location and suggest using `+=` or `-=`, or setting `enable_vector_component_overwrites=True` when direct reassignment is needed.

@Ramzee-S, if this answers your question, please feel free to close this issue. The new issue can cover the diagnostic work.

**Comment by Ramzee-S** [comment]
Thanks @shi-eric for the quick, clear answer — that fully resolves it. The restriction is indeed documented in the differentiability guide; we missed that paragraph, apologies. The explanation of why the copy-based lowering is needed in the general nonlinear case is appreciated, and the planned opt-in diagnostic sounds like exactly the right addition. Closing as answered — thank you!
